subsys/pm/device.c

```diff
@@ -292,3 +292,19 @@ bool pm_device_on_power_domain(const struct device *dev)
 	return false;
 #endif
 }
+
+bool pm_device_is_powered(const struct device *dev)
+{
+#ifdef CONFIG_PM_DEVICE_POWER_DOMAIN
+	struct pm_device *pm = dev->pm;
+
+	if (pm == NULL) {
+		/* We can't tell without access to the PM struct, assume the device is powered */
+		return true;
+	}
+	return (pm->domain == NULL) ||
+	       (pm->domain->pm->state == PM_DEVICE_STATE_ACTIVE);
```

Comment:
the device may be suspended (it is up to devices decided what to do when suspended) and not belongs to a power domain.

Also, I'm wondering about if having an API like this is better than keep the device state consistent and just return it.

Reply:
This helper is about the physical power state of the device, not some internal powered/not powered value.

The primary use case I have for this is inside of device init functions, when you need to know whether the device is powered before you can attempt to talk to it. At that point the device state isn't good enough, because you're the one thats supposed to be setting it up.